world-chain-builder/crates/world/payload/src/builder.rs

```diff
@@ -643,8 +625,35 @@ where
                 return Ok(Some(()));
             }
 
+            // If the transaction is verified, check if it can be added within the verified gas limit
+            if let Some(payloads) = pooled_tx.pbh_payload() {
+                if info.cumulative_gas_used + tx.gas_limit() > verified_gas_limit {
+                    best_txs.mark_invalid(tx.signer(), tx.nonce());
+                    continue;
+                }
+
+                if payloads.iter().any(|payload| {
+                    let dup = spent_nullifier_hashes.contains(&payload.nullifier_hash);
+                    spent_nullifier_hashes.insert(payload.nullifier_hash);
+                    dup
+                }) {
+                    best_txs.mark_invalid(tx.signer(), tx.nonce());
+                    invalid_txs.push(*pooled_tx.hash());
+                    continue;
+                }
```

Comment:
Since we are using a `HashSet` this can be further simplified to the following:


```rust
if payloads.iter().any(|payload| !spent_nullifier_hashes.insert(payload.nullifier_hash)) {
    best_txs.mark_invalid(tx.signer(), tx.nonce());
    invalid_txs.push(*pooled_tx.hash());
    continue;
}
```